### DISCUSSION TITLE
Unexpected Behavior of x-for with null Value

### DISCUSSION THREAD

**Discussion by koojunho** [alpinejs/alpine#4318] [1. Help] [answered]
Hello,

I encountered an issue with the following code:

```
<div x-data="{}">
  <template x-for="v of null">
    <span x-text="v"></span>
  </template>
</div>
```

When rendered in the browser, it shows 1. Why does null behave like an array with one item? How can I prevent this behavior?

Thank you!

**Comment by SimoTod** [comment, marked as answer]
It's because JS is weird and `isNaN(null)` returns false even if null is not really a number: https://github.com/alpinejs/alpine/blob/main/packages/alpinejs/src/directives/x-for.js#L295C42-L295C47

so alpine treats it as a range loop and then [this line](https://github.com/alpinejs/alpine/blob/main/packages/alpinejs/src/directives/x-for.js#L51C21-L51C64) with null returns [1] since in javascript `Array(null)` returns `[null]` (but `Array(0)` returns `[]`).

you can probably send a PR to fix the isNumeric function in that file and add a test. You'll also need to update https://github.com/alpinejs/alpine/blob/main/packages/alpinejs/src/directives/x-for.js#L54 to return the empty array for null otherwise `Object.entries(null)` will error further down in the code

**Reply by SimoTod** [reply]
If you don't want to fix the bug and you are looking for a userland workaround, you can do stuff like
```html
<div x-data="{foo: {bar: 'bar'}, baz: null}">
  <template x-for="v of (foo ?? {})">
    <span x-text="v"></span>
  </template>
  <template x-for="v of (baz ?? {})">
    <span x-text="v"></span>
  </template>
  <template x-for="v in (baz ?? [])">
    <span x-text="v"></span>
  </template>
</div>
```

**Reply by ekwoka** [reply]
brb, PRing a fix to JavaScript

**Reply by ekwoka** [reply]
And technically, while Alpine doesn't care about `of` or `in`, `of` is for array items, and `in` is for object keys.

**Reply by koojunho** [reply]
Thank you for the response. I'll try using x-for="v in (baz ?? [])".

**Reply by ekwoka** [reply]
Ideally, just never let it be null.

It's not typesafe.

If it's empty/unpopulated, make it be `[]` to start.
